### ISSUE TITLE
Error: Cannot read property 'navigator' of undefined

### ISSUE THREAD

**Issue by NehaSharma2304** [highcharts/highcharts#8091]
I wanted to display export icon on highcharts
exporting: {
			enabled: true
		},

This code doesn't work. Therefore, I wrote following lines 

****const Highcharts = require('highcharts');
require ('highcharts/modules/exporting.js')(Highcharts);**** 

But I am getting,

/Users/nsharma/nodejs/node_modules/highcharts/modules/exporting.js:9
[2] (function(k){"object"===typeof module&&module.exports?module.exports=k:k(Highcharts)})(function(k){(function(f){var k=f.defaultOptions,p=f.doc,B=f.Chart,x=f.addEvent,I=f.removeEvent,F=f.fireEvent,r=f.createElement,C=f.discardElement,v=f.css,n=f.merge,D=f.pick,h=f.each,G=f.objectEach,t=f.extend,J=f.isTouchDevice,E=f.win,H=E.navigator.userAgent,K=f.Renderer.prototype.symbols;/Edge\/|Trident\/|MSIE /.test(H);/firefox/i.test(H);t(k.lang,{printChart:"Print chart",downloadPNG:"Download PNG image",downloadJPEG:"Download JPEG image",
[2]                                                                                                                                                                                                                                                                                                                                     ^
[2] 
[2] TypeError: Cannot read property 'navigator' of undefined
[2]     at /Users/nsharma/nodejs/node_modules/highcharts/modules/exporting.js:9:325
[2]     at /Users/nsharma/nodejs/node_modules/highcharts/modules/exporting.js:27:356
[2]     at Object.<anonymous> (/Users/nsharma/nodejs/public/components/DataChart/datachart.js:9:1)
[2]     at Module._compile (module.js:570:32)
[2]     at loader (/Users/nsharma/nodejs/node_modules/babel-register/lib/node.js:144:5)
[2]     at Object.require.extensions.(anonymous function) [as .js] (/Users/nsharma/nodejs/node_modules/babel-register/lib/node.js:154:7)
[2]     at Module.load (module.js:487:32)
[2]     at tryModuleLoad (module.js:446:12)
[2]     at Function.Module._load (module.js:438:3)
[2]     at Module.require (module.js:497:17)
[2] [nodemon] app crashed - waiting for file changes before starting...

Please help.

**Comment by pawelfus** [comment]
Hi @NehaSharma2304 

Thank you for reporting the issue. Could you provide more details? 
- node and npm versions
- Highcharts version
- datachart.js (part with chart initialization and Highcharts import)

Thanks!

**Comment by KacperMadej** [comment]
Hi @NehaSharma2304 

You wrote that you are using Highcharts 5.0.7, but from the error is looks like you are not using matching exporting module - it should be loaded from the files of Highcharts in the same version. The navigator is checked since 5.0.12 in exporting module and was improved (bug resolved) later in version 6.0.0. 

Please try with Highcharts in version 6+ and with modules of the same code version, because of the issue https://github.com/highcharts/highcharts/issues/7031

BTW: We have an official wrapper for React: https://github.com/highcharts/highcharts-react/

**Comment by NehaSharma2304** [comment]
Hi, thanks for the reply, I have updated Highchart version to  "highcharts": "^6.0.7" but still this error is coming.

**Comment by KacperMadej** [comment]
@NehaSharma2304 
We'll need to investigate the case further - please send use the files required to recreate the issue or recreate the problem in an online demo - https://codesandbox.io/

For better communication you could contact us using e-mail: support@highcharts.com

**Comment by donal-crotty** [comment]
Hi, experiencing the same issue as @NehaSharma2304 . Has a solution to this been found? Would really like to add exporting to my chart.

**Comment by pawelfus** [comment]
Hi @donal-crotty 

We still couldn't recreate issue, could you follow Kacper's instructions (https://github.com/highcharts/highcharts/issues/8091#issuecomment-378223452) and let us know? Thanks!

**Comment by NehaSharma2304** [comment]
Hi @donal-crotty, pardon me for not being active in reply. I was able to implement this feature using react-csv npm package.

**Comment by sturmenta** [comment]
What I did was:

- Do not use the latest version of firebase, but the [version that indicate that you use](https://firebase.google.com/docs/firestore/quickstart?hl=es-419)
`npm install firebase@4.12.1 --save`

- Replace

```javascript
import * firebase from 'firebase';
import '@firebase/firestore';
import '@firebase/messaging';
```

by

```javascript
import * firebase from 'firebase';
import 'firebase/firestore';
import 'firebase/messaging';
```

- `npm i -S @firebase/firestore` `npm i -S @firebase/messaging`

And everything worked correctly!

**Comment by pawelfus** [comment]
Hi @PadharaHitesh 

Unfortunately, this is not enough to let us debug the issue and fix it, please provide more details. See https://github.com/highcharts/highcharts/issues/8091#issuecomment-416837853

**Comment by Yaseen-Nadaf** [comment]
Hi, I am getting the same error as @NehaSharma2304, But I am using highcharts with angular universal, client-side rendering is all good but when it comes to server-side rendering, it's giving me that error.

Here are my configs:

Angular CLI: 6.0.0
Node: 8.11.3
OS: win32 x64
Angular: 6.1.10
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, platform-server, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.6.0
@angular-devkit/build-angular     0.6.0
@angular-devkit/build-optimizer   0.6.0
@angular-devkit/core              0.6.0
@angular-devkit/schematics        0.6.0
@angular/cdk                      6.4.7
@angular/cli                      6.0.0
@angular/flex-layout              6.0.0-beta.18
@angular/material                 6.4.7
@ngtools/webpack                  6.0.0
@schematics/angular               0.6.0
@schematics/update                0.6.0
rxjs                              6.3.3
typescript                        2.9.2
webpack                           4.6.0

Following are highcharts versions from package.json:

"highcharts": "^6.1.4",
"highcharts-angular": "^2.3.1",

Please tell me if you need any other info for debugging this.

**Comment by Yaseen-Nadaf** [comment]
This is my code for chart:

```import { Component, OnInit } from '@angular/core';
import * as Highcharts from 'highcharts';
import * as HC_exporting from 'highcharts/modules/exporting';
HC_exporting(Highcharts);

@Component({
  selector: 'app-data',
  templateUrl: './data.component.html',
  styleUrls: ['./data.component.css']
})
export class DataComponent implements OnInit {
  Highcharts = Highcharts; // required
  chartConstructor = 'chart'; // optional string, defaults to 'chart'
  chartOptions = {
    series: [{
      type: 'column',
      data: [1, 2, 3]
    }]
  }; // required

  updateFlag = false; // optional boolean
  oneToOneFlag = true; // optional boolean, defaults to false
  runOutsideAngularFlag = false; // optional boolean, defaults to false


  constructor() { }

  ngOnInit() {
  }
}

**Comment by KacperMadej** [comment]
@Yaseen-Nadaf Looks like you are using highcharts-angular wrapper, so please add a live demo and report the issue on the wrapper's GitHub issues page. https://github.com/highcharts/highcharts-angular/

**Comment by fazelmk** [comment]
Having the ```TypeError: Cannot read property 'navigator' of undefined``` in exporting module too.
In my case it's caused by webWorker. Worked fine before I add webWorker.

Any ideas on how to get working with webWorker?

I'm using Angular 5 and Highcharts 6.2.0. 
And I'm not using the angular wrapper.

**Comment by alexliu-github** [comment]
I have the same error return when I run the code locally, using latest react, highcharts and highcharts-react-official

import Highcharts from 'highcharts'
import exporting from 'highcharts/modules/exporting'
import HighchartsReact from 'highcharts-react-official'
exporting(Highcharts)  // App is crashed trigger by exporting function

>  starting `node build/server.js`
> C:\Development\fcui10-2\node_modules\highcharts\modules\exporting.js:9
> (function(g){"object"===typeof module&&module.exports?module.exports=g:"function"===typeof define&&define.amd?define(function(){return g}):g(Highcharts)})(function(g){(function(f){var g=f.defaultOptions,y=f.doc,z=f.Chart,q=f.addEvent,H=f.removeEvent,B=f.fireEvent,r=f.createElement,C=f.discardElement,u=f.css,p=f.merge,A=f.pick,l=f.each,E=f.objectEach,t=f.extend,I=f.isTouchDevice,D=f.win,F=D.navigator.userAgent,J=f.Renderer.prototype.symbols;/Edge\/|Trident\/|MSIE /.test(F);/firefox/i.test(F);t(g.lang,
> ^
> 
> TypeError: Cannot read property 'navigator' of undefined
> at C:\Development\fcui10-2\node_modules\highcharts\modules\exporting.js:9:394
> at C:\Development\fcui10-2\node_modules\highcharts\modules\exporting.js:28:32

**Comment by KacperMadej** [comment]
@fazelmk @alexliu-github 
This error could be seen when exporting module doesn't have the same version as Highcharts. Please try loading .src.js instead of .js files to use not minified code and if you could create a live demo then it would be great. Version of Highcharts could be checked using Highcharts.version, but for modules the info about version is only in a comment at the beginning of each file.

Another source of the problem could be that Highcharts are loaded in an environment that doesn't have standard browser `window` variable defined and current context (`this`) for Highcharts initialization doesn't have `navigator` property the same as standard `window`'s `navigator` property. In this case you could check if such behavior could be resolved in Angular / React / Node environment - e.g. load Highcharts with on default browser's `window` as a context.

**Comment by fazelmk** [comment]
@KacperMadej In my case I think it's just the lack of a DOM. As webWorkers execute in a different process then the DOM the exporting module don't have access to it and therefore raise the error.

**Comment by alexliu-github** [comment]
**@KacperMadej** 
Thank you for your explanation, I understand what caused the error message, I am able to get it working and move forward.

**Comment by KacperMadej** [comment]
@fazelmk In such case you could load module (`import * as HC_exporting from 'highcharts/modules/exporting';`), but postpone initialization (`HC_exporting(Highcharts);`) until DOM is available or do both, load and init, later when it will have DOM accessible.

**Comment by MaxFlower** [comment]
> @KacperMadej In my case I think it's just the lack of a DOM. As webWorkers execute in a different process then the DOM the exporting module don't have access to it and therefore raise the error.

Yes, you're right. I use Angular and this problem was solved when I put `HC_exporting(Highcharts)` into `ngAfterViewInit` function.
